pep-0497-externals.txt

```diff
@@ -0,0 +1,359 @@
+PEP: 497
+Title: External and environment-specific dependencies
+Version: $Revision$
+Last-Modified: $Date$
+Author: Tennessee Leeuwenburg <tleeuwenburg@gmail.com>,
+
+Status: Draft
+Type: Standards Track
+Python-Version 3.6
+Content-Type: text/x-rst
+Created: 03-Aug-2015
+Post-History: 03-Aug-2015
+
+
+Abstract
+========
+
+The canonical method for installing Python packages is to use pip, whether that
+is referencing PyPI or a local repository. A limitation of the current install
+ecosystem is the relationship with non-Python dependencies. This
+can result in compilation failures or execution failures. Package maintainers
+may wish to specify the requirements to avoid this issue, and users may
+appreciate earlier and more informative messages about their system.
+
+The responsibility of pip is to stop early and provide useful information about
+missing (and not automatically installable) dependencies. Ideally it would provide
+a useful user message about how to obtain the required dependencies. The
+responsibility for actual installation lies with the user at this time.
+
+This PEP proposes a mechanism for extending the requirements section of setup.py
+which will allow authors to express non-Python requirements cleanly. In the
+implementation section, a discussion of how this may be used to solved problems
+is given.
+
+Here is an example of the proposed syntax, where both operating system and build
+target specifiers are included. This is "Concept 1" of the alternatives discussed
+later in the document.
+
+::
+
+    setup(
+        name='uses_scipy_with_atlas_extension'
+        install_requires=['yaml', 'scipy[atlas]', 'lib!libblas1.so'],
+        setup_requires=[
+            'pyfortran',
+            'bin!fortran:dist_name=="ubuntu"',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+    )
+
+
+If the user does not have the required dependencies, the user will be presented
+with a useful install message and a process that exits early, as opposed to
+relying on the compiler to fail and provide an error message that may not be
+readily understood by most users.
+
+The mechanism of referencing failed dependencies such as 'kernel.h' to useful
+user messages regarding the actual operating system packages that are relevant
+is discussed in the implementation section, along with the rationale for
+handling that referencing outside of setup.py.
+
```

Comment:
We need to add in here a scoping thing.

This PEP does not (and shouldn't - its a bigger and different problem) address the problem of choosing a particular wheel. It *does* cover the problem of detecting whether a given wheel that you've already chosen is able to be installed and run.

So something like:
"In summary, this PEP covers:
 - specifying external build and run time dependencies for source trees, sdists and (run time only) wheels
 - metadata for describing external build and runtime dependencies

This PEP does not cover:
 - identifying which wheel is likely to be suitable to install on a given Linux platform. That problem, while related, is separate.